### ISSUE TITLE
Sphinx for docstring translation and processing?

### ISSUE THREAD

**Issue by timcera** [rubik/mando#21]
Sphinx can translate Google and Numpy formatted docstrings into reStructuredText format.  See http://www.sphinx-doc.org/en/stable/ext/napoleon.html.  I think mando could also transfer other parts of the docstring processing of the reStructuredText to Sphinx.

If you think it would be a good idea then:
    Should Sphinx be a prerequisite?
OR
    Should we just take the code we need from Sphinx and keep with mando.

If Sphinx is a prerequisite then mando could take advantage of any improvements or bug fixes, but it is a pretty hefty library to carry around.  I always have it installed anyway, but it is unlikely to be installed for most users (or is it?).  This would be the easiest approach - but you would have to accept the idea of having a prerequisite.

I have started taking from Sphinx and including just the Google/Numpy translation feature into mando.  Lots of problems with this approach because there are many connections to other parts of the Sphinx ecosystem, and you have to track them all down. Also someone has to take care of that code once it is brought over.  Perhaps not often, but it might mean folding in complicated changes or bug fixes from Sphinx.  Before I put any more work into it however, I wanted to run the idea past you.

**Comment by rubik** [comment]
It's probably a good idea to leverage Sphinx's power for docstring processing. I don't think requiring it would be a problem, since with package managers the dependency handling is streamlined. And, as you noted, adapting the code would be much more difficult and I might add, a frail approach.

I'll have to take a look at Sphinx and see what parts could be helpful, do you already have an idea?

**Comment by timcera** [comment]
This have been quite a learning experience.  Sphinx doesn't have what I expected it to have.

Sphinx only works with file systems.  You can use sphinx.ext.napoleon to translate Google and Numpy docstrings to ReST, but very difficult to use anything else.

In fact, Sphinx works intimately with the file system.  You specify the source directory, the configuration directory (that holds "conf.py") and the output directory.  There doesn't seem to be a way to just pass a ReST docstring to a function and have it rendered.

"sphinxify.py" is part of "sage" at https://github.com/sagemath/sagenb/blob/master/sagenb/misc/sphinxify.py.  He creates a temporary directory and configuration file to run Sphinx.  Very clunky.  Also really closely tied to other "sage" libraries.

"rst2html" is part of "docutils" which is the foundational library under Sphinx.  However "docutils" does not have a "rst2ascii".  Sphinx does have a text writer which I thought could work, but it is setup differently than the "docutils" writers and I couldn't figure out a way for "docutils" to accept the Sphinx text writer.

"restview" creates HTML then spawns a web-browser.  Not even close.

So then I found "rst2ansi" at https://pypi.python.org/pypi/rst2ansi/.  This is looking pretty good.  Cons are that it was only Python 3.3+ (until he accepts my pull request to include support for 2.7) and it uses docutils to render.  Using docutils means that the :param var: and :type int: tags aren't rendered since those are Sphinx extensions.

Pygments is also a possibility, though doesn't render as nice as "rst2ansi".

    from pygments.lexers.markup import RstLexer
    from pygments.formatters import TerminalFormatter
    code = open('plotdoc.rst', 'r').read()
    print(pygments.highlight(code, RstLexer(), TerminalFormatter()))

Kindest regards,
Tim

**Comment by rubik** [comment]
Oh, I thought Sphinx had some kind of API to interface with it programmatically. Very good research work! It seems that rst2ansi is our only option then. What about `:param:`? Will it remain in the output? I'm don't know exactly, but maybe Pygments can parse it.

**Comment by timcera** [comment]
In my test with Pygments, it didn't handle the :param var: directive.

I think mando would have to parse out the :param var: and :type int: directives and replace with pure ReST to use "rst2ansi" or Pygments.

**Comment by rubik** [comment]
Mhm but then what's the advantage of using rst2ansi? In my opinion it would have been best to delegate the parsing to a specialized library in order to support more cases and in a more robust manner. Since Sphinx turned out to be unusable in this regard, then why should we adopt rst2ansi if it does not support Sphinx's ReST?

**Comment by rubik** [comment]
Fixed by #24.

**Comment by timcera** [comment]
#24 actually didn't address this.  I see this as a different issue than converting Google and Numpy docstirngs to ReST and then adapting that ReST format to mando's.

This would allow mando to not worry about indentation, wrapping, tables, ...etc. by giving the barely processed ReST docstring to Sphinx (if an API can be found) or rst2ansi.

If there is a Sphinx API, then mando would do very little, process the docstring for it's own purposes, and give the entire docstring to Sphinx for rendering.

If mando uses rst2ansi would have to make a format change:

Sphinx:
```
:param teststr: This is the test string.
:type teststr: str
```
rst2ansi:
```
teststr: str
    This is a test string formatted as a docutils 
    definition list.
```

**Comment by rubik** [comment]
I don't think it's wise to change the format in order to support `rst2ansi`. By keeping the Sphinx format, it's possible to generate the CLI *and* the package documentation.

**Comment by rubik** [comment]
This can now be closed since it has been implemented. However, Sphinx' dependencies have grown too much to be required by mando, and we don't even need them, except for the very small napoleon dependency. This is the current dependency tree:
```
  - mando
    - sphinx
      - alabaster
      - babel
        - pytz
      - docutils
      - imagesize
      - Jinja2
        - MarkupSafe
      - Pygments
      - requests
        - certifi
        - chardet
        - idna
        - urllib3
      - setuptools
      - six
      - snowballstemmer
      - sphinxcontrib-websupport
```

If we included napoleon as a vendored dependency, this would be the result:
```
  - mando
    - sphinx
      - alabaster
      - babel
        - pytz
      - docutils
      - imagesize
      - Jinja2
        - MarkupSafe
      - Pygments
      - requests
        - certifi
        - chardet
        - idna
        - urllib3
      - setuptools
      - six
```

The Sphinx dependency is the only one (apart from the optional `rst2ansi`) and 99% of it is never used by mando. Therefore I've now decided to remove the Sphinx dependency and include the napoleon extension in mando as a vendored dependency. This is addressed by #42.
